flake.nix

```diff
@@ -0,0 +1,140 @@
+# Copyright (c) 2024, Amazon.com, Inc. or its affiliates. All rights reserved.
+#
+# See LICENSE for licensing information
+
+{
+  description = "aws-ofi-nccl development/build flake.";
+
+  outputs =
+    { self, flake-parts, ... }@inputs:
+    let
+      inherit (inputs.lib-aggregate) lib;
+      systems = [
+        "x86_64-linux"
+        "aarch64-linux"
+      ];
+    in
+    flake-parts.lib.mkFlake { inherit inputs; } (
+      { withSystem, flake-parts-lib, ... }:
+      {
+        inherit systems;
+        imports = [
+          inputs.git-hooks.flakeModule
+          flake-parts.flakeModules.easyOverlay
+        ];
+        flake = {
+          githubActionChecks = inputs.nix-github-actions.lib.mkGithubMatrix {
+            checks = self.outputs.packages.x86_64-linux;
+          };
+        };
+        debug = true;
+        perSystem =
+          {
+            system,
+            config,
+            final,
+            pkgs,
+            ...
+          }:
+          {
+            _module.args.pkgs = import inputs.nixpkgs {
+              inherit system;
+              overlays = [
+                (import ./nix/overlays/libfabric)
+                inputs.cuda-packages.overlays.default
+                inputs.self.overlays.default
+              ];
+              config = {
+                cudaSupport = true;
+                cudaForwardCompat = true;
+                cudaCapabilities = [
+                  "7.0"
+                  "7.5"
+                  "8.0"
+                  "8.6"
+                  "8.9"
+                  "9.0"
+                  "9.0a"
+                ];
+                allowBroken = true;
+                allowUnfree = true;
+              };
+            };
+            pre-commit.settings = import ./nix/checks.nix { inherit lib; };
+            devShells.default = import ./nix/shell.nix {
+              inherit
+                pkgs
+                config
+                system
+                inputs
+                self
+                ;
+            };
+            overlayAttrs = {
+              cudaPackagesExtensions = [
+                (import ./nix/cudaPackagesExtensions/0001-add-latest-nccl.nix { inherit (pkgs) fetchFromGitHub; })
+                (import ./nix/cudaPackagesExtensions/0002-use-latest-nccl.nix)
+                (import ./nix/cudaPackagesExtensions/0003-nccl-tests-use-mpi.nix { inherit config; })
+                (import ./nix/cudaPackagesExtensions/0004-add-ncclAws.nix {
+                  inherit lib config;
+                  inherit (pkgs) symlinkJoin patchelf;
+                })
+                (import ./nix/cudaPackagesExtensions/0005-add-nccl-tests-aws.nix {
+                  inherit (pkgs) replaceDependency;
+                })
```

Comment:
Why not push the arguments inside the extension? That is, instead of passing `pkgs.fetchFromGitHub` to the extension, you should be able to have something like

```nix
ffinal:
let 
 inherit (ffinal.pkgs) fetchFromGitHub;
in
pprev:
...
```

`pkgs` is a member of every `cudaPackages` package set; this would also prevent your `pkgs` from contaminating the overlays generated by `overlayAttrs` (because you're passing things from your `pkgs` to them directly instead of allowing the extension to retrieve them from whatever fixed point it is being evaluated within).

I'd recommend changing usages in the extensions of `config.packages.blarg` with `ffinal.pkgs.blarg` for the same reason; if you're using overlays properly, everything that you'd normally need to pass through directly will be available in the resulting Nixpkgs instantiation which is used to evaluate these extensions.